src/mutation/RelayMutation.js

```diff
@@ -50,12 +50,29 @@ class RelayMutation<Tp: Object> {
     route: RelayMetaRoute
   ) => Variables;
 
-  props: Tp;
+  _context: RelayContext;
   _didShowFakeDataWarning: boolean;
+  _unresolvedProps: Tp;
+  props: Tp;
 
   constructor(props: Tp) {
-    this._didShowFakeDataWarning = false;
-    this._resolveProps(props);
+    this._unresolvedProps = props;
+  }
+
+  /**
+   * @internal
+   */
+  bindContext(context: RelayContext): void {
+    if (!this._context) {
+      this._context = context;
+      this._resolveProps();
```

Comment:
Existing products may have logic in mutation constructors that depends on `this.props` being set. With this new change, there isn't a clear place to put that logic. I wonder if we should add something like an optional "did receive props" hook that would provide a clear way to upgrade mutations. We could probably even write a codemod that would move existing constructor logic after the `super` call into the callback method.

cc @wincent